### PR TITLE
Remove `async` from `describe` blocks and add missing `await` in tests

### PR DESCRIPTION
**Pull request by fvictorio**
When testing the latest version of Hardhat in the OZ repo, we noticed that some of the tests were suddenly failing. The problem was caused by a combination of some missing or extra `async`/`await` keywords. Mainly: the functions of some `describe` blocks had the `async` keyword (and they shouldn't), and some async assertions were missing the `await` keyword. This, combined with the different performance characteristics of the latest HH version, surfaced some issues that weren't showing up before just by coincidence. I fixed that in fe64f5c8.

After that, some other tests started failing. Most of them are related to event assertions that either had the wrong order of arguments, or had a missing argument. These were actually failing before, but since they weren't being awaited, they were floating rejected promises that didn't make the tests fail. I fixed that in 258a203d.

Now there is only one test failing, the `grants role to the user` test in `AccessManager.test.js`. This is asserting that a `RoleAccessRequested` was emitted but, as far as I can tell, that event doesn't exist anymore. I couldn't find an equivalent one so I just left it as it was; I'm sure you can fix it faster than me.

#### PR Checklist

This PR only affects tests, so no docs change or changeset entry is needed.

- [x] Tests
- [ ] Documentation
- [ ] Changeset entry (run `npx changeset add`)

### CODE REVIEW

**Comment by fvictorio** [comment]
I forgot to mention:

- The wrong `async` keywords in describe functions can be prevented with [this one](https://github.com/lo1tuma/eslint-plugin-mocha/blob/main/docs/rules/no-async-describe.md).
- The missing `await` keywords are harder. Ideally you'd use typescript and the `no-floating-promises` rule, but I understand that that would be a huge undertaking.

**Comment by ernestognw** [comment]
Hey @fvictorio, thank you for raising this. After we migrated tests to Ethers I think this was one of our major worries (getting ghost tests).

For the record, this comes from [this PR](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4710/files#diff-29eec4d6198c2ac37cf692929fdc77636999e93b1580ec73f34d418567e24408L1369-R1256). I guess the oversight came from the previous order of the arguments but I can't explain how `RoleAccessRequested` appeared.

I fixed the tests, and I think we'll need to discuss adding the `no-async-describe` rule (@Amxx ). Unfortunately for the `await` cases, I don't think we're supporting Typescript soon, not only is it a big commitment but is increasing the dependency surface, and the benefit comes only for testing purposes.

**Review by ernestognw — APPROVED** [review]
LGTM, we need to discuss the options to avoid it happening again, but the PR is good to go rn.

**Comment by fvictorio** [comment]
Thanks for the review!

**Comment by fvictorio** [comment]
Maybe you can consider adding this somewhere in the tests setup:

```js
process.on("unhandledRejection", (reason, promise) => {
  throw promise;
});
```

**Comment by Amxx** [comment]
Changes to the test files look good. 

Is the hardhat update necessary, it comes with a big `package-lock.json` diff ...

**Comment by ernestognw** [comment]
@fvictorio we see a lot of changes to the package-lock.json, is there any reason for upgrading Hardhat?

We may do this in a future PR but I wonder if we need it now

**Comment by fvictorio** [comment]
No, no need to update it now. You might get a 10-15% performance improvement, but there isn't a strong reason for you to upgrade. I can rebase that commit out if you want.

**Comment by ernestognw** [comment]
> No, no need to update it now. You might get a 10-15% performance improvement, but there isn't a strong reason for you to upgrade. I can rebase that commit out if you want.

Yeah I think we'd prefer to keep this PR short. We still want to upgrade for next hardfork though

**Comment by fvictorio** [comment]
Done!

File: hardhat/env-artifacts.js
Comment (ernestognw) [inline hardhat/env-artifacts.js:4]: I couldn't find anything useful in the second argument of the listener (the promise). It doesn't give the stacktrace anyway and this one works fine.

The only drawback is that it halts the tests whenever a failure is found. We can tweak it to behave differently but I'm personally fine with it.

**Review by ernestognw — APPROVED** [review]
Thanks for all your work @fvictorio 🙏🏻 
I created an issue for the ESLint rule so we can discuss it later: https://github.com/OpenZeppelin/openzeppelin-contracts/issues/4943

Rn it's looking good again, let's wait on @Amxx approval.

File: hardhat/env-artifacts.js
Comment (Amxx) [reply hardhat/env-artifacts.js:4]: I wouldn't have put it in this file. IMO this should be its own file in `hardhat/`.

Maybe `hardhat/test-sanity.js`

File: test/proxy/beacon/BeaconProxy.test.js
Comment (Amxx) [inline test/proxy/beacon/BeaconProxy.test.js:37]: I guess the issue here is a `ERC1967InvalidImplementation` with address(1). We should test that.

File: test/token/ERC1155/extensions/ERC1155Supply.test.js
Comment (Amxx) [inline test/token/ERC1155/extensions/ERC1155Supply.test.js:114]: This is missing an await, no?

Comment (Amxx) [reply test/token/ERC1155/extensions/ERC1155Supply.test.js:114]:
```suggestion
      await this.token.$_update(ethers.ZeroAddress, ethers.ZeroAddress, [firstTokenId], [firstTokenValue]);
```

File: test/proxy/beacon/BeaconProxy.test.js
Comment (Amxx) [reply test/proxy/beacon/BeaconProxy.test.js:37]:
```suggestion
      await expect(this.newBeaconProxy(badBeacon, '0x'))
        .to.be.revertedWithCustomError(this.factory, 'ERC1967InvalidImplementation');
```

Comment (Amxx) [inline test/proxy/beacon/BeaconProxy.test.js:37]: Here the revert is:
- BadBeaconNoImpl does not implement `implementation()`
- ERC1967 utils try to get that in `_setBeacon`
- the call fail (no impelmentation of the selector + no fallback)

**Comment by Amxx** [comment]
Moved 
```javascript
process.on('unhandledRejection', reason => {
  throw new Error(reason);
});
```
to a dedicated file + added a missing await (are there any more?)

File: test/proxy/beacon/BeaconProxy.test.js
Comment (ernestognw) [reply test/proxy/beacon/BeaconProxy.test.js:39]: It was previously:

```
await expect(this.newBeaconProxy(badBeacon, '0x')).to.be.revertedWithoutReason
```

Which seemed correct to me but I'm realizing it was missing the parenthesis:

```
await expect(this.newBeaconProxy(badBeacon, '0x')).to.be.revertedWithoutReason();
```

We can keep it with the explicit `revertedWithoutReason()` and keep the comments.

File: hardhat/env-artifacts.js
Comment (ernestognw) [reply hardhat/env-artifacts.js]: I'm fine with `test-sanity.js`

**Comment by ernestognw** [comment]
> to a dedicated file + added a missing await (are there any more?)

Right, the one you spotted is a tricky one given is expecting no changes. The test should make sure the function is called before asserting, though. I don't expect other tests to be in similar cases.

**Review by ernestognw — APPROVED** [review]

**Review by Amxx — APPROVED** [review]

**Pull request merged**
